client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,814 @@
+				return 0
+			case change.EffectiveNumber() > k.effectiveNumber && change.canonHeight > k.signalBlockNumber:
+				return 1
+			case change.EffectiveNumber() < k.effectiveNumber && change.canonHeight < k.signalBlockNumber:
+				return -1
+			default:
+				panic("huh?")
+			}
+		},
+	)
+
+	logger.Debugf(
+		"inserting potential forced set hashNumber at block number %d (delayed by %d blocks).",
+		pending.canonHeight, pending.delay,
+	)
+
+	// Insert hashNumber at index
+	if len(authSet.pendingForcedChanges) == idx {
+		authSet.pendingForcedChanges = append(authSet.pendingForcedChanges, pending)
+	} else if len(authSet.pendingForcedChanges) > idx {
+		authSet.pendingForcedChanges = append(
+			authSet.pendingForcedChanges[:idx+1], authSet.pendingForcedChanges[idx:]...)
+		authSet.pendingForcedChanges[idx] = pending
+	} else {
+		panic("invalid insertion into pending forced changes")
+	}
+
+	logger.Debugf(
+		"there are now %d pending forced changes",
+		len(authSet.pendingForcedChanges),
+	)
+
+	return nil
+}
+
+// addPendingChange Note an upcoming pending transition. Multiple pending standard changes
+// on the same branch can be added as long as they don't overlap. Forced
+// changes are restricted to one per fork. This method assumes that changes
+// on the same branch will be added in-order. The given function
+// `is_descendent_of` should return `true` if the second hash (target) is a
+// descendent of the first hash (base).
+func (authSet *AuthoritySet[H, N]) addPendingChange(pending PendingChange[H, N], isDescendentOf IsDescendentOf[H]) error {
+	if InvalidAuthorityList(pending.nextAuthorities) {
+		return errInvalidAuthoritySet
+	}
+
+	switch pending.delayKind.value.(type) {
+	case Finalized:
+		return authSet.addStandardChange(pending, isDescendentOf)
+	case Best[N]:
+		return authSet.addForcedChange(pending, isDescendentOf)
+	default:
+		panic("delayKind is invalid type")
+	}
+}
+
+// pendingChanges Inspect pending changes. Standard pending changes are iterated first,
+// and the changes in the roots are traversed in pre-order, afterwards all
+// forced changes are iterated.
+func (authSet *AuthoritySet[H, N]) pendingChanges() []PendingChange[H, N] {
+	// get everything from standard hashNumber roots
+	changes := authSet.pendingStandardChanges.PendingChanges()
+
+	// append forced changes
+	changes = append(changes, authSet.pendingForcedChanges...)
+
+	return changes
+}
+
+// CurrentLimit Get the earliest limit-block number, if any. If there are pending changes across
+// different forks, this method will return the earliest effective number (across the
+// different branches) that is higher or equal to the given min number.
+//
+// Only standard changes are taken into account for the current
+// limit, since any existing forced hashNumber should preclude the voter from voting.
+func (authSet *AuthoritySet[H, N]) CurrentLimit(min N) (limit *N) {
```

Comment:
This is `pub(crate)` in [substrate](https://github.com/paritytech/substrate/blob/7c8b89b3a6b603902709fc33fdb10a864b8b4204/client/consensus/grandpa/src/authorities.rs#L423).
```suggestion
func (authSet *AuthoritySet[H, N]) currentLimit(min N) (limit *N) {
```